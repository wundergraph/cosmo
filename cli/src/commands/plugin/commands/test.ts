import { Command } from 'commander';
import { resolve } from 'pathe';
import pc from 'picocolors';
import Spinner from 'ora';
import { BaseCommandOptions } from '../../../core/types/types.js';
import { checkAndInstallTools, runGoTests } from '../toolchain.js';

export default (opts: BaseCommandOptions) => {
  const command = new Command('test');
  command.description('Run tests for a gRPC router plugin');
  command.argument('[directory]', 'Directory of the plugin', '.');
  command.option('-l, --language <language>', 'Programming language of the plugin', 'go');
  command.action(async (directory, options) => {
    const startTime = performance.now();
    const pluginDir = resolve(directory);
    const spinner = Spinner({ text: 'Running tests...' });

    try {
      spinner.start();

      if (options.language.toLowerCase() !== 'go') {
        spinner.warn(pc.yellow(`Language '${options.language}' is not supported yet. Using 'go' instead.`));
        options.language = 'go';
      }

      // Check and install tools if needed
      if (!options.skipToolsInstallation) {
        await checkAndInstallTools(options.forceToolsInstallation);
      }

      const srcDir = resolve(pluginDir, 'src');

      spinner.text = 'Running tests...';

      const { failed } = await runGoTests(srcDir, spinner);

      // Calculate elapsed time
      const endTime = performance.now();
      const elapsedTimeMs = endTime - startTime;
      const formattedTime =
        elapsedTimeMs > 1000 ? `${(elapsedTimeMs / 1000).toFixed(2)}s` : `${Math.round(elapsedTimeMs)}ms`;

      if (failed) {
        spinner.fail(pc.red(`Tests failed! [${formattedTime}]`));
      } else {
        spinner.succeed(pc.green(`Tests completed successfully! [${formattedTime}]`));
      }
    } catch (error: any) {
      spinner.fail(pc.red(`Failed to run tests: ${error.message}`));
      throw error;
    } finally {
      spinner.stop();
    }
  });

  return command;
};
