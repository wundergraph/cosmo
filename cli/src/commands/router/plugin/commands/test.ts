import path from 'node:path';
import os from 'node:os';
import { Command } from 'commander';
import { resolve } from 'pathe';
import Spinner from 'ora';
import { BaseCommandOptions } from '../../../../core/types/types.js';
import { checkAndInstallTools, installGoDependencies, runGoTests } from '../toolchain.js';
import { renderResultTree } from '../helper.js';

export default (opts: BaseCommandOptions) => {
  const command = new Command('test');
  command.description('Run tests for a gRPC router plugin');
  command.argument('[directory]', 'Directory of the plugin', '.');
  command.option('--skip-tools-installation', 'Skip tool installation', false);
  command.option(
    '--force-tools-installation',
    'Force tools installation regardless of version check or confirmation',
    false,
  );
  command.action(async (directory, options) => {
    const startTime = performance.now();
    const pluginDir = resolve(directory);
    const spinner = Spinner({ text: 'Running tests...' });
    const pluginName = path.basename(pluginDir);

    try {
      spinner.start();

      // Check and install tools if needed
      if (!options.skipToolsInstallation) {
        await checkAndInstallTools(options.forceToolsInstallation);
      }

      spinner.text = 'Installing Go dependencies...';

      await installGoDependencies(pluginDir, spinner);

      const srcDir = resolve(pluginDir, 'src');

      spinner.text = 'Running tests...';

      try {
        const { failed } = await runGoTests(srcDir, spinner, false);

        // Calculate elapsed time
        const endTime = performance.now();
        const elapsedTimeMs = endTime - startTime;
        const formattedTime =
          elapsedTimeMs > 1000 ? `${(elapsedTimeMs / 1000).toFixed(2)}s` : `${Math.round(elapsedTimeMs)}ms`;

        // Common details for both success and failure
        const details = {
          source: srcDir,
          env: `${os.platform()} ${os.arch()}`,
          time: formattedTime,
        };

        const title = failed ? 'Tests failed!' : 'Tests completed successfully!';
        renderResultTree(spinner, title, !failed, pluginName, details);
      } catch (error: any) {
        renderResultTree(spinner, 'Tests failed!', false, pluginName, {
          source: srcDir,
          env: `${os.platform()} ${os.arch()}`,
          error: error.message,
        });
      }
    } catch (error: any) {
      renderResultTree(spinner, 'Failed to run tests!', false, pluginName, {
        env: `${os.platform()} ${os.arch()}`,
        error: error.message,
      });
    } finally {
      spinner.stop();
    }
  });

  return command;
};
