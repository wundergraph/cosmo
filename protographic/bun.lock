{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "protographic",
      "dependencies": {
        "graphql": "^16.11.0",
        "lodash": "^4.17.21",
      },
      "devDependencies": {
        "@types/bun": "latest",
        "@types/lodash": "^4.17.16",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.2.11", "", { "dependencies": { "bun-types": "1.2.11" } }, "sha512-ZLbbI91EmmGwlWTRWuV6J19IUiUC5YQ3TCEuSHI3usIP75kuoA8/0PVF+LTrbEnVc8JIhpElWOxv1ocI1fJBbw=="],

    "@types/lodash": ["@types/lodash@4.17.16", "", {}, "sha512-HX7Em5NYQAXKW+1T+FiuG27NGwzJfCX3s1GjOa7ujxZa52kjJLOr4FUxT+giF6Tgxv1e+/czV/iTtBw27WTU9g=="],

    "@types/node": ["@types/node@22.15.3", "", { "dependencies": { "undici-types": "~6.21.0" } }, "sha512-lX7HFZeHf4QG/J7tBZqrCAXwz9J5RD56Y6MpP0eJkka8p+K0RY/yBTW7CYFJ4VGCclxqOLKmiGP5juQc6MKgcw=="],

    "bun-types": ["bun-types@1.2.11", "", { "dependencies": { "@types/node": "*" } }, "sha512-dbkp5Lo8HDrXkLrONm6bk+yiiYQSntvFUzQp0v3pzTAsXk6FtgVMjdQ+lzFNVAmQFUkPQZ3WMZqH5tTo+Dp/IA=="],

    "graphql": ["graphql@16.11.0", "", {}, "sha512-mS1lbMsxgQj6hge1XZ6p7GPhbrtFwUFYi3wRzXAC/FmYnyXMTvvI3td3rjmQ2u8ewXueaSvRPWaEcgVVOT9Jnw=="],

    "lodash": ["lodash@4.17.21", "", {}, "sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "undici-types": ["undici-types@6.21.0", "", {}, "sha512-iwDZqg0QAGrg9Rav5H4n0M64c3mkR59cJ6wQp+7C4nI0gsmExaedaYLNO44eT4AtBBwjbTiGPMlt2Md0T9H9JQ=="],
  }
}
