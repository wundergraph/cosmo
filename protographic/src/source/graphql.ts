import {
  Kind,
  parse,
  type DefinitionNode,
  type FieldDefinitionNode,
  type InputObjectTypeDefinitionNode,
  type InputValueDefinitionNode,
  type ObjectTypeDefinitionNode,
  type TypeNode,
} from "graphql";
import {
  type Message,
  type RPC,
  type Field,
  type Service,
  ServiceBuilder,
  type Type,
} from "@/common/service";

export function intoIntermediate(serviceName: string, sdl: string): Service {
  const ast = parse(sdl);

  const svcBuilder = new ServiceBuilder(serviceName);

  for (const node of ast.definitions) {
    switch (node.kind) {
      case Kind.OBJECT_TYPE_DEFINITION:
        switch (node.name.value) {
          case "Subscription":
          case "Mutation":
          case "Query":
            svcBuilder.addRPCs(...operationObjectToRPCs(node));
            break;
          default:
            svcBuilder.addMessages(objectToMessage(node));
            break;
        }

        break;
      case Kind.INPUT_OBJECT_TYPE_DEFINITION:
        svcBuilder.addMessages(objectToMessage(node));
        break;
      default:
        console.error("unsupported graphql type used: " + node.kind);
    }
  }

  return svcBuilder.build();
}

function operationObjectToRPCs(node: ObjectTypeDefinitionNode): RPC[] {
  const rpcs: RPC[] = [];

  for (const field of node.fields ?? []) {
    rpcs.push(operationFieldToRPC(node.name.value, field));
  }

  return rpcs;
}

function operationFieldToRPC(type: string, node: FieldDefinitionNode): RPC {
  const rpc: RPC = {
    name: node.name.value,
    kind: "Query",
    type: typeFromTypeNode(node.type),
    arguments: [],
  };

  switch (type) {
    case "Subscription":
      rpc.kind = "Subscription";
      break;
    case "Mutation":
      rpc.kind = "Mutation";
      break;
    case "Query":
      rpc.kind = "Query";
      break;
    default:
      console.error("unsupported graphql type used: " + type);
  }

  for (const field of node.arguments ?? []) {
    rpc.arguments.push(fieldNodeToField(field));
  }

  return rpc;
}

function fieldNodeToField(
  input: InputValueDefinitionNode | FieldDefinitionNode,
): Field {
  const field: Field = {
    name: input.name.value,
    type: typeFromTypeNode(input.type),

    resolved: false,
  };

  return field;
}

function typeFromTypeNode(type: TypeNode): Type {
  const typeObj: Type = { name: "", required: false, list: false };

  let fieldType = type;

  if (fieldType.kind === Kind.NON_NULL_TYPE) {
    typeObj.required = true;
    fieldType = fieldType.type;
  }

  if (fieldType.kind === Kind.LIST_TYPE) {
    typeObj.list = true;
    if (fieldType.type.kind === Kind.NON_NULL_TYPE) {
      fieldType = fieldType.type.type;
    } else {
      fieldType = fieldType.type;
    }
  }

  if (fieldType.kind === Kind.NAMED_TYPE) {
    typeObj.name = fieldType.name.value;
  }

  return typeObj;
}

function objectToMessage(
  node: ObjectTypeDefinitionNode | InputObjectTypeDefinitionNode,
): Message {
  const message: Message = {
    name: node.name.value,
    fields: [],
  };

  for (const field of node.fields ?? []) {
    message.fields.push(fieldNodeToField(field));
  }

  return message;
}
