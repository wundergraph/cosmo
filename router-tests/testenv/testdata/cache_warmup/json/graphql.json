{
  "query": "query { employees { id details { forename } } }"
}