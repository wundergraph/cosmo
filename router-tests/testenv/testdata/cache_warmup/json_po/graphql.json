{
  "request": {
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f"
      }
    }
  },
  "client": {
    "name": "my-client",
    "version": "missing"
  }
}