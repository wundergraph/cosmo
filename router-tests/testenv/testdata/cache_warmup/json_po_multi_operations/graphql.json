{
  "request": {
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "724399f210ef3f16e6e5427a70bb9609ecea7297e99c3e9241d5912d04eabe60"
      }
    }
  },
  "client": {
    "name": "my-client",
    "version": "missing"
  }
}