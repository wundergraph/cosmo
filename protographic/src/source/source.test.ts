import { describe, it, expect } from "bun:test";
import { intoIntermediate } from "./graphql";

describe("GraphQL source", () => {
  const sdl = `
    scalar ID
    scalar String
    scalar Float

    type Product @key(fields: "id") {
      id: ID!
      name: String!
      price: Float!
      # shippingEstimate(input: ShippingEstimateInput!): Float!
    }

    type Storage @key(fields: "id") {
      id: ID!
      name: String!
      location: String!
    }

    type User {
      id: ID!
      name: String!
    }

    type NestedTypeA {
      id: ID!
      name: String!
      b: NestedTypeB!
    }

    type NestedTypeB {
      id: ID!
      name: String!
      c: NestedTypeC!
    }

    type NestedTypeC {
      id: ID!
      name: String!
    }

    type RecursiveType {
      id: ID!
      name: String!
      recursiveType: RecursiveType!
    }

    type TypeWithMultipleFilterFields {
      id: ID!
      name: String!
      filterField1: String!
      filterField2: String!
    }

    input FilterTypeInput {
      filterField1: String!
      filterField2: String!
    }

    type TypeWithComplexFilterInput {
      id: ID!
      name: String!
    }

    input FilterType {
      name: String!
      filterField1: String!
      filterField2: String!
      pagination: Pagination
    }

    input Pagination {
      page: Int!
      perPage: Int!
    }

    input ComplexFilterTypeInput {
      filter: FilterType!
    }

    type Query {
      _entities(representations: [_Any!]!): [_Entity!]!
      users: [User!]!
      user(id: ID!): User!
      userIdOptional(id: ID): User
      nestedType: [NestedTypeA!]!
      recursiveType: RecursiveType!
      typeFilterWithArguments(
        filterField1: String!
        filterField2: String!
      ): [TypeWithMultipleFilterFields!]!
      typeWithMultipleFilterFields(
        filter: FilterTypeInput!
      ): [TypeWithMultipleFilterFields!]!
      complexFilterType(
        filter: ComplexFilterTypeInput!
      ): [TypeWithComplexFilterInput!]!
    }

    type Mutation {
      createProduct(input: CreateProductInput!): Product!
    }

    input CreateProductInput {
      name: String!
      price: Float!
    }

    union _Entity = Product | Storage
    scalar _Any
  `;

  it("should correctly read GraphQL SDL into intermediate format", () => {
    const result = intoIntermediate("ProductService", sdl);

    expect(result).toMatchSnapshot();
  });
});
