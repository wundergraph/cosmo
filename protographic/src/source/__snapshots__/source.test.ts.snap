// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL source should correctly read GraphQL SDL into intermediate format 1`] = `
{
  "messages": [
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "price",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Float",
            "required": true,
          },
        },
      ],
      "name": "Product",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "location",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "name": "Storage",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "name": "User",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "b",
          "resolved": false,
          "type": {
            "list": false,
            "name": "NestedTypeB",
            "required": true,
          },
        },
      ],
      "name": "NestedTypeA",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "c",
          "resolved": false,
          "type": {
            "list": false,
            "name": "NestedTypeC",
            "required": true,
          },
        },
      ],
      "name": "NestedTypeB",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "name": "NestedTypeC",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "recursiveType",
          "resolved": false,
          "type": {
            "list": false,
            "name": "RecursiveType",
            "required": true,
          },
        },
      ],
      "name": "RecursiveType",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "filterField1",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "filterField2",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "name": "TypeWithMultipleFilterFields",
    },
    {
      "fields": [
        {
          "name": "filterField1",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "filterField2",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "name": "FilterTypeInput",
    },
    {
      "fields": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "name": "TypeWithComplexFilterInput",
    },
    {
      "fields": [
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "filterField1",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "filterField2",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "pagination",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Pagination",
            "required": false,
          },
        },
      ],
      "name": "FilterType",
    },
    {
      "fields": [
        {
          "name": "page",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Int",
            "required": true,
          },
        },
        {
          "name": "perPage",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Int",
            "required": true,
          },
        },
      ],
      "name": "Pagination",
    },
    {
      "fields": [
        {
          "name": "filter",
          "resolved": false,
          "type": {
            "list": false,
            "name": "FilterType",
            "required": true,
          },
        },
      ],
      "name": "ComplexFilterTypeInput",
    },
    {
      "fields": [
        {
          "name": "name",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "price",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Float",
            "required": true,
          },
        },
      ],
      "name": "CreateProductInput",
    },
  ],
  "name": "ProductService",
  "rpcs": [
    {
      "arguments": [
        {
          "name": "representations",
          "resolved": false,
          "type": {
            "list": true,
            "name": "_Any",
            "required": true,
          },
        },
      ],
      "kind": "Query",
      "name": "_entities",
      "type": {
        "list": true,
        "name": "_Entity",
        "required": true,
      },
    },
    {
      "arguments": [],
      "kind": "Query",
      "name": "users",
      "type": {
        "list": true,
        "name": "User",
        "required": true,
      },
    },
    {
      "arguments": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
      ],
      "kind": "Query",
      "name": "user",
      "type": {
        "list": false,
        "name": "User",
        "required": true,
      },
    },
    {
      "arguments": [
        {
          "name": "id",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": false,
          },
        },
      ],
      "kind": "Query",
      "name": "userIdOptional",
      "type": {
        "list": false,
        "name": "User",
        "required": false,
      },
    },
    {
      "arguments": [],
      "kind": "Query",
      "name": "nestedType",
      "type": {
        "list": true,
        "name": "NestedTypeA",
        "required": true,
      },
    },
    {
      "arguments": [],
      "kind": "Query",
      "name": "recursiveType",
      "type": {
        "list": false,
        "name": "RecursiveType",
        "required": true,
      },
    },
    {
      "arguments": [
        {
          "name": "filterField1",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
        {
          "name": "filterField2",
          "resolved": false,
          "type": {
            "list": false,
            "name": "String",
            "required": true,
          },
        },
      ],
      "kind": "Query",
      "name": "typeFilterWithArguments",
      "type": {
        "list": true,
        "name": "TypeWithMultipleFilterFields",
        "required": true,
      },
    },
    {
      "arguments": [
        {
          "name": "filter",
          "resolved": false,
          "type": {
            "list": false,
            "name": "FilterTypeInput",
            "required": true,
          },
        },
      ],
      "kind": "Query",
      "name": "typeWithMultipleFilterFields",
      "type": {
        "list": true,
        "name": "TypeWithMultipleFilterFields",
        "required": true,
      },
    },
    {
      "arguments": [
        {
          "name": "filter",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ComplexFilterTypeInput",
            "required": true,
          },
        },
      ],
      "kind": "Query",
      "name": "complexFilterType",
      "type": {
        "list": true,
        "name": "TypeWithComplexFilterInput",
        "required": true,
      },
    },
    {
      "arguments": [
        {
          "name": "input",
          "resolved": false,
          "type": {
            "list": false,
            "name": "CreateProductInput",
            "required": true,
          },
        },
      ],
      "kind": "Mutation",
      "name": "createProduct",
      "type": {
        "list": false,
        "name": "Product",
        "required": true,
      },
    },
  ],
}
`;
